File: network-monitor/Aggregator.cpp
#include "Aggregator.hpp"

#include <algorithm>
#include <iostream>

using namespace std;

void Aggregator::update(const ResolvedPacket& packet) {
    lock_guard<mutex> lock(mtx_);

    auto& p = process_stats_[packet.pid];
    p.pid = packet.pid;
    p.process_name = packet.process_name;
    p.user = packet.user;
    p.bytes += packet.length;
    p.packets += 1;

    auto& u = user_stats_[packet.user];
    u.user = packet.user;
    u.bytes += packet.length;
    u.packets += 1;

    auto& pr = protocol_stats_[packet.protocol];
    pr.protocol = packet.protocol;
    pr.bytes += packet.length;
    pr.packets += 1;
}

bool Aggregator::has_data() const {
    lock_guard<mutex> lock(mtx_);
    return !process_stats_.empty() || !user_stats_.empty() || !protocol_stats_.empty();
}

void Aggregator::print_summary() const {
    vector<ProcAgg> procs;
    vector<UserAgg> users;
    vector<ProtoAgg> protos;

    {
        lock_guard<mutex> lock(mtx_);

        if (process_stats_.empty() && user_stats_.empty() && protocol_stats_.empty()) {
            return;
        }

        for (const auto& [pid, stat] : process_stats_) {
            procs.push_back(stat);
        }

        for (const auto& [user, stat] : user_stats_) {
            users.push_back(stat);
        }

        for (const auto& [proto, stat] : protocol_stats_) {
            protos.push_back(stat);
        }
    }

    sort(procs.begin(), procs.end(),
              [](const ProcAgg& a, const ProcAgg& b) {
                  return a.bytes > b.bytes;
              });

    sort(users.begin(), users.end(),
              [](const UserAgg& a, const UserAgg& b) {
                  return a.bytes > b.bytes;
              });

    sort(protos.begin(), protos.end(),
              [](const ProtoAgg& a, const ProtoAgg& b) {
                  return a.bytes > b.bytes;
              });

    cerr << "\n========== Final Aggregation Summary ==========\n";

    cerr << "Top Processes:\n";
    for (size_t i = 0; i < min<size_t>(5, procs.size()); ++i) {
        cerr << "  PID=" << procs[i].pid
                  << "  PROC=" << procs[i].process_name
                  << "  USER=" << procs[i].user
                  << "  BYTES=" << procs[i].bytes
                  << "  PACKETS=" << procs[i].packets
                  << "\n";
    }

    cerr << "Top Users:\n";
    for (size_t i = 0; i < min<size_t>(5, users.size()); ++i) {
        cerr << "  USER=" << users[i].user
                  << "  BYTES=" << users[i].bytes
                  << "  PACKETS=" << users[i].packets
                  << "\n";
    }

    cerr << "Protocols:\n";
    for (const auto& proto : protos) {
        cerr << "  PROTO=" << proto.protocol
                  << "  BYTES=" << proto.bytes
                  << "  PACKETS=" << proto.packets
                  << "\n";
    }

    cerr << "===============================================\n";
}

File: rust_monitor/src/aggregator.rs
use std::collections::BTreeMap;

use crate::packet_types::ResolvedPacket;

#[derive(Debug, Clone)]
pub struct ProcAgg {
    pub pid: i32,
    pub process_name: String,
    pub user: String,
    pub bytes: u64,
    pub packets: u64,
}

#[derive(Debug, Clone)]
pub struct UserAgg {
    pub user: String,
    pub bytes: u64,
    pub packets: u64,
}

#[derive(Debug, Clone)]
pub struct ProtoAgg {
    pub protocol: String,
    pub bytes: u64,
    pub packets: u64,
}

#[derive(Debug, Default)]
pub struct Aggregator {
    process_stats: BTreeMap<i32, ProcAgg>,
    user_stats: BTreeMap<String, UserAgg>,
    protocol_stats: BTreeMap<String, ProtoAgg>,
}

impl Aggregator {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn update(&mut self, packet: &ResolvedPacket) {
        let p = self.process_stats.entry(packet.pid).or_insert_with(|| ProcAgg {
            pid: packet.pid,
            process_name: packet.process_name.clone(),
            user: packet.user.clone(),
            bytes: 0,
            packets: 0,
        });
        p.process_name = packet.process_name.clone();
        p.user = packet.user.clone();
        p.bytes += packet.length;
        p.packets += 1;

        let u = self.user_stats.entry(packet.user.clone()).or_insert_with(|| UserAgg {
            user: packet.user.clone(),
            bytes: 0,
            packets: 0,
        });
        u.bytes += packet.length;
        u.packets += 1;

        let pr = self.protocol_stats.entry(packet.protocol.clone()).or_insert_with(|| ProtoAgg {
            protocol: packet.protocol.clone(),
            bytes: 0,
            packets: 0,
        });
        pr.bytes += packet.length;
        pr.packets += 1;
    }

    pub fn has_data(&self) -> bool {
        !(self.process_stats.is_empty() && self.user_stats.is_empty() && self.protocol_stats.is_empty())
    }

    pub fn print_summary(&self) {
        if !self.has_data() {
            return;
        }

        let mut procs: Vec<_> = self.process_stats.values().cloned().collect();
        let mut users: Vec<_> = self.user_stats.values().cloned().collect();
        let mut protos: Vec<_> = self.protocol_stats.values().cloned().collect();

        procs.sort_by(|a, b| b.bytes.cmp(&a.bytes));
        users.sort_by(|a, b| b.bytes.cmp(&a.bytes));
        protos.sort_by(|a, b| b.bytes.cmp(&a.bytes));

        eprintln!("\n========== Final Aggregation Summary ==========");
        eprintln!("Top Processes:");
        for p in procs.iter().take(5) {
            eprintln!(
                "  PID={}  PROC={}  USER={}  BYTES={}  PACKETS={}",
                p.pid, p.process_name, p.user, p.bytes, p.packets
            );
        }

        eprintln!("Top Users:");
        for u in users.iter().take(5) {
            eprintln!(
                "  USER={}  BYTES={}  PACKETS={}",
                u.user, u.bytes, u.packets
            );
        }

        eprintln!("Protocols:");
        for pr in &protos {
            eprintln!(
                "  PROTO={}  BYTES={}  PACKETS={}",
                pr.protocol, pr.bytes, pr.packets
            );
        }
        eprintln!("===============================================");
    }
}
